File: matiasmasso/Mat.NET/DT2/DTOECITRansmGroup.vb
Public Class DTOECITransmGroup

    Inherits DTOBaseGuid

    Property Ord As Integer
    Property Nom As String
    Property Platform As DTOContact

    Property Items As List(Of DTOECITransmCentre)

    Public Sub New()
        MyBase.New()
        _Items = New List(Of DTOECITransmCentre)
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
        _Items = New List(Of DTOECITransmCentre)
    End Sub

    Shared Function SelectedDeliveries(oGroup As DTOECITransmGroup, oSrc As List(Of DTODelivery), ByRef rest As List(Of DTODelivery)) As List(Of DTODelivery)
        Dim retval As New List(Of DTODelivery)
        rest = New List(Of DTODelivery)
        For Each oDelivery As DTODelivery In oSrc
            If DTOECITransmGroup.Belongs(oGroup, oDelivery) Then
                retval.Add(oDelivery)
            Else
                rest.Add(oDelivery)
            End If
        Next
        Return retval
    End Function


    Shared Function SortDeliveries(oDeliveries As List(Of DTODelivery)) As List(Of DTODelivery)
        Dim retval As List(Of DTODelivery) = oDeliveries.
            OrderBy(Function(x) x.Platform.Guid.ToString()).
            OrderBy(Function(x) x.Customer.Ref).
            OrderBy(Function(x) DTOEci.NumeroDeComanda(x)).
            ToList
        Return retval
    End Function

    Shared Function Belongs(oGroup As DTOECITransmGroup, oDelivery As DTODelivery) As Boolean
        Dim retval As Boolean = oGroup.MatchesPlatform(oDelivery) And oGroup.MatchesCenter(oDelivery)
        Return retval
    End Function
    Shared Function Belongs(oGroup As DTOECITransmGroup, oOrder As DTOPurchaseOrder) As Boolean
        Dim retval As Boolean = oGroup.MatchesPlatform(oOrder) And oGroup.MatchesCenter(oOrder)
        Return retval
    End Function

    Public Function MatchesPlatform(oDelivery As DTODelivery) As Boolean
        Dim retval = _Platform.Equals(oDelivery.platform)
        Return retval
    End Function

    Public Function MatchesPlatform(oOrder As DTOPurchaseOrder) As Boolean
        Dim retval = _Platform.Equals(oOrder.platform)
        Return retval
    End Function

    Public Function MatchesCenter(oDelivery As DTODelivery) As Boolean
        Dim retval As Boolean
        If _Items.Count = 0 Then
            retval = True 'si no hi ha centres, s'admet qualsevol centre
        Else
            retval = _Items.Any(Function(x) x.Centre.Equals(oDelivery.customer))
        End If
        Return retval
    End Function

    Public Function MatchesCenter(oOrder As DTOPurchaseOrder) As Boolean
        Dim retval As Boolean
        If _Items.Count = 0 Then
            retval = True 'si no hi ha centres, s'admet qualsevol centre
        Else
            retval = _Items.Any(Function(x) x.Centre.Equals(oOrder.customer))
        End If
        Return retval
    End Function

    Shared Function Contacts(oECITransmGroup As DTOECITransmGroup) As List(Of DTOContact)
        Dim retval As New List(Of DTOContact)
        For Each item In oECITransmGroup.Items
            retval.Add(item.Centre)
        Next
        Return retval
    End Function

    Shared Function Sort(oDeliveries As List(Of DTODelivery)) As List(Of DTODelivery)
        Dim retval As List(Of DTODelivery) = oDeliveries.
            OrderBy(Function(x) x.Platform.Guid.ToString()).
            OrderBy(Function(x) x.Customer.Ref).
            OrderBy(Function(x) DTOEci.NumeroDeComanda(x)).
            ToList
        Return retval
    End Function

    Shared Function SortedDeliveries(oGroups As List(Of DTOECITransmGroup), oDeliveries As List(Of DTODelivery), exs As List(Of Exception)) As List(Of DTODelivery)
        Dim retval As New List(Of DTODelivery)
        Dim rest As List(Of DTODelivery) = oDeliveries
        For Each oGroup As DTOECITransmGroup In oGroups
            If rest.Count = 0 Then Exit For

            Dim src As List(Of DTODelivery) = rest
            Dim oGroupDeliveries As List(Of DTODelivery) = DTOECITransmGroup.SelectedDeliveries(oGroup, src, rest)
            Dim oSortedGroupDeliveries = oGroupDeliveries.
                OrderBy(Function(x) x.customer.ref).
                OrderBy(Function(x) DTOEci.NumeroDeComanda(x)).
                ToList
            retval.AddRange(oSortedGroupDeliveries)
        Next

        If rest.Count > 0 Then
            Dim sb As New Text.StringBuilder
            sb.AppendLine(String.Format("{0} albarans no s'han pogut assignar a cap grup de transmisió:", rest.Count))
            For Each oDelivery In rest
                sb.AppendLine(String.Format("centre {0} comanda {1}", oDelivery.customer.ref, DTOEci.NumeroDeComanda(oDelivery)))
            Next
        End If
        Return retval
    End Function


End Class

File: matiasmasso/Mat.NET/DTO/DTOECITransmGroup.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DTO
{
    public class DTOECITransmGroup : DTOBaseGuid
    {
        public int Ord { get; set; }
        public string Nom { get; set; }
        public DTOContact Platform { get; set; }

        public List<DTOECITransmCentre> Items { get; set; }

        public DTOECITransmGroup() : base()
        {
            Items = new List<DTOECITransmCentre>();
        }

        public DTOECITransmGroup(Guid oGuid) : base(oGuid)
        {
            Items = new List<DTOECITransmCentre>();
        }

        public static List<DTODelivery> SelectedDeliveries(DTOECITransmGroup oGroup, List<DTODelivery> oSrc, ref List<DTODelivery> rest)
        {
            List<DTODelivery> retval = new List<DTODelivery>();
            rest = new List<DTODelivery>();
            foreach (DTODelivery oDelivery in oSrc)
            {
                if (DTOECITransmGroup.Belongs(oGroup, oDelivery))
                    retval.Add(oDelivery);
                else
                    rest.Add(oDelivery);
            }
            return retval;
        }


        public static List<DTODelivery> SortDeliveries(List<DTODelivery> oDeliveries)
        {
            List<DTODelivery> retval = oDeliveries.OrderBy(x => x.Platform.Guid.ToString()).OrderBy(x => x.Customer.Ref).OrderBy(x => DTOEci.NumeroDeComanda(x)).ToList();
            return retval;
        }

        public static bool Belongs(DTOECITransmGroup oGroup, DTODelivery oDelivery)
        {
            bool retval = oGroup.MatchesPlatform(oDelivery) & oGroup.MatchesCenter(oDelivery);
            return retval;
        }
        public static bool Belongs(DTOECITransmGroup oGroup, DTOPurchaseOrder oOrder)
        {
            bool retval = oGroup.MatchesPlatform(oOrder) & oGroup.MatchesCenter(oOrder);
            return retval;
        }

        public bool MatchesPlatform(DTODelivery oDelivery)
        {
            var retval = Platform.Equals(oDelivery.Platform);
            return retval;
        }

        public bool MatchesPlatform(DTOPurchaseOrder oOrder)
        {
            var retval = Platform.Equals(oOrder.Platform);
            return retval;
        }

        public bool MatchesCenter(DTODelivery oDelivery)
        {
            bool retval;
            if (Items.Count == 0)
                retval = true; // si no hi ha centres, s'admet qualsevol centre
            else
                retval = Items.Any(x => x.Centre.Equals(oDelivery.Customer));
            return retval;
        }

        public bool MatchesCenter(DTOPurchaseOrder oOrder)
        {
            bool retval;
            if (Items.Count == 0)
                retval = true; // si no hi ha centres, s'admet qualsevol centre
            else
                retval = Items.Any(x => x.Centre.Equals(oOrder.Customer));
            return retval;
        }

        public static List<DTOContact> Contacts(DTOECITransmGroup oECITransmGroup)
        {
            List<DTOContact> retval = new List<DTOContact>();
            foreach (var item in oECITransmGroup.Items)
                retval.Add(item.Centre);
            return retval;
        }

        public static List<DTODelivery> Sort(List<DTODelivery> oDeliveries)
        {
            List<DTODelivery> retval = oDeliveries.OrderBy(x => x.Platform.Guid.ToString()).OrderBy(x => x.Customer.Ref).OrderBy(x => DTOEci.NumeroDeComanda(x)).ToList();
            return retval;
        }

        public static List<DTODelivery> SortedDeliveries(List<DTOECITransmGroup> oGroups, List<DTODelivery> oDeliveries, List<Exception> exs)
        {
            List<DTODelivery> retval = new List<DTODelivery>();
            List<DTODelivery> rest = oDeliveries;
            foreach (DTOECITransmGroup oGroup in oGroups)
            {
                if (rest.Count == 0)
                    break;

                List<DTODelivery> src = rest;
                List<DTODelivery> oGroupDeliveries = DTOECITransmGroup.SelectedDeliveries(oGroup, src, ref rest);
                var oSortedGroupDeliveries = oGroupDeliveries.OrderBy(x => x.Customer.Ref).OrderBy(x => DTOEci.NumeroDeComanda(x)).ToList();
                retval.AddRange(oSortedGroupDeliveries);
            }

            if (rest.Count > 0)
            {
                System.Text.StringBuilder sb = new System.Text.StringBuilder();
                sb.AppendLine(string.Format("{0} albarans no s'han pogut assignar a cap grup de transmisió:", rest.Count));
                foreach (var oDelivery in rest)
                    sb.AppendLine(string.Format("centre {0} comanda {1}", oDelivery.Customer.Ref, DTOEci.NumeroDeComanda(oDelivery)));
            }
            return retval;
        }

        public override string ToString()
        {
            return Nom;
        }
    }
}
